### PR TITLE
fix: move initialization of P1 smart contract metrics

### LINKED ISSUE #16902: fix: move initialization of P1 smart contract metrics
P1 smart contract metrics aren't initialized correctly. Move the initialization code.  The new place for it has the potential to be called multiple times but metrics must only be created once.  So there needs to be a gate too.

### PR DESCRIPTION
**Pull request by david-bakin-sl**
**Description**:

Move initialization call of P1 smart contract metrics.

The new place for it has the potential to be called multiple times but metrics must only be created once. So there is a gate too.

**Related issue(s)**:

Fixes #16902 

**Checklist**

- [x] Documented (Code comments, README, etc.)
- [x[ Tested (unit, integration, etc.)

### CODE REVIEW

File: hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/metrics/ContractMetrics.java [hunk 1]
```diff
@@ -88,61 +87,72 @@ public ContractMetrics(
...
+        // remarkably, that can happen more than once.  We must only create the metrics the
+        // _first_ time.  Hence, the simple gate here.
+
+        synchronized (metricsCreationLock) {
+            if (!primaryMetricsAreCreated) {
+
+                final var contractsConfig = requireNonNull(contractsConfigSupplier.get());
+                this.p1MetricsEnabled = contractsConfig.metricsSmartContractPrimaryEnabled();
+
+                final var metrics = requireNonNull(metricsSupplier.get());
```
Comment (lukelee-sl) [inline hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/metrics/ContractMetrics.java:106]: Can this be inside the if statement below?

**Review by lukelee-sl — APPROVED** [review]
LGTM

File: hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/metrics/ContractMetrics.java [hunk 2]
```diff
@@ -88,61 +87,72 @@ public ContractMetrics(
...
-                final var name = toRejectedName(REJECTED_TYPE3_FUNCTIONALITY, REJECTED_TYPE3_SHORT_DESCR);
-                final var descr = toRejectedDescr(REJECTED_TYPE3_FUNCTIONALITY, REJECTED_TYPE3_SHORT_DESCR);
-                final var config = new Counter.Config(METRIC_CATEGORY, name)
-                        .withDescription(descr)
-                        .withUnit(METRIC_TXN_UNIT);
-                final var metric = newCounter(metrics, config);
-                rejectedEthType3Counter = metric;
+        // Primary metrics are a fixed set and can be created when `Hedera` initializes the system.
+        // But it actually must wait until the platform calls `Hedera.onStateInitialized` and,
+        // remarkably, that can happen more than once.  We must only create the metrics the
```
Comment (tinker-michaelj) [inline hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/metrics/ContractMetrics.java:97]: There is a pretty mundane explanation for this; if a node's network connection is interrupted and then restored, its reconnect phase will happen within the same JVM; and will hence re-initialize the same `Hedera` singleton.

Comment (david-bakin-sl) [reply hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/metrics/ContractMetrics.java:97]: yes, i finally figured that out.  but in fact, in the P2 metrics PR (coming shortly) I figured out the _correct_ way to handle this using InitTrigger.GENESIS, so all this garbage is out.

File: hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/metrics/ContractMetrics.java [hunk 3]
```diff
@@ -88,61 +87,72 @@ public ContractMetrics(
...
+        // _first_ time.  Hence, the simple gate here.
+
+        synchronized (metricsCreationLock) {
```
Comment (tinker-michaelj) [inline hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/metrics/ContractMetrics.java:100]: This shouldn't require any synchronization.

**Comment by david-bakin-sl** [comment]
> I wonder if we should support unregistering metrics so the lifecycle in this PR would look more like the "unregister-any-registered-listeners-from-before-reconnect" pattern [here](https://github.com/hashgraph/hedera-services/blob/develop/hedera-node/hedera-app/src/main/java/com/hedera/node/app/Hedera.java#L982).

Perhaps if the underlying metrics implementation (provided by the platform side) needed that sort of thing.

Note that the platform's `Metrics` class _does_ support removing metrics but the only use (outside of test cases for it) is to remove a per-node uptime metric for a node that has disappeared from the roster.

And ofc we're not currently doing it - e.g., the relatively new `TssMetrics` doesn't have such a feature.

File: hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/metrics/ContractMetrics.java (hunk 3 above)
Comment (david-bakin-sl) [reply hedera-node/hedera-smart-contract-service-impl/src/main/java/com/hedera/node/app/service/contract/impl/exec/metrics/ContractMetrics.java:100]: yes, and in the follow on PR it doesn't - as I ensure it is only called if the trigger condition is GENESIS.

**Pull request merged**